<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.280">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Dmitry Kondrashov">

<title>Intro to data (lab 1 for BIOS 26122)</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="lab1_intro_sol_files/libs/clipboard/clipboard.min.js"></script>
<script src="lab1_intro_sol_files/libs/quarto-html/quarto.js"></script>
<script src="lab1_intro_sol_files/libs/quarto-html/popper.min.js"></script>
<script src="lab1_intro_sol_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="lab1_intro_sol_files/libs/quarto-html/anchor.min.js"></script>
<link href="lab1_intro_sol_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="lab1_intro_sol_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="lab1_intro_sol_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="lab1_intro_sol_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="lab1_intro_sol_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Intro to data (lab 1 for BIOS 26122)</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Dmitry Kondrashov </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<section id="description" class="level2">
<h2 class="anchored" data-anchor-id="description">Description</h2>
<p>The goal of this assignment is to introduce basic data description and visualization using base R and tidyverse packages. Specifically, you should be able to:</p>
<ol type="1">
<li>Read in a data set as a data frame (or tibble) and report the number and types of variables, as well as the number of observations.</li>
<li>Visualize a variable using a method appropriate for its type; observe the general shape of the distribution.</li>
<li>Report descriptive statistics (e.g.&nbsp;counts, means, standard deviations) for numeric variables and generate cross tables for categorical variables.</li>
<li>Perform an assessment of missing values and outliers and do basic data cleaning.</li>
</ol>
<p>These are the first steps that one needs to do to begin working with a data set. Below you will see several real data sets to practice on.</p>
</section>
<section id="heart-rates" class="level2">
<h2 class="anchored" data-anchor-id="heart-rates">Heart rates</h2>
<p>The following data set contains heart rates measured and reported by students in my class Introduction to Quantitative Modeling for Biology. There are four different heart rates measured (two at rest and two after exercise) and the year it was measured.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>heart_rates <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"https://raw.githubusercontent.com/dkon1/intro-ml-bio/main/labs/data/HR_data_combined.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 1068 Columns: 5
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
dbl (5): Rest1, Ex1, Rest2, Ex2, Year

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
</div>
<ol type="1">
<li><p>Report the number and types of variables and the number of observations</p>
<p>There are 5 variables (columns), of which 4 are numeric and the 5th (year) is (arguably) categorical (even though it’s a number, it’s not meant for calculations).</p></li>
<li><p>Make a histogram of one of the numeric variables of your choice; describe its shape (e.g.&nbsp;normal, uniform, bimodal) and comment on any interesting features.</p></li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># base R:</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(heart_rates<span class="sc">$</span>Ex2, <span class="at">main =</span> <span class="st">'Histogram of Ex2'</span>, <span class="at">xlab =</span> <span class="st">"heart rate (bpm)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab1_intro_sol_files/figure-html/unnamed-chunk-2-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ggplot</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>heart_rates <span class="sc">|&gt;</span> <span class="fu">ggplot</span>() <span class="sc">+</span> <span class="fu">aes</span>(<span class="at">x=</span>Ex2) <span class="sc">+</span> <span class="fu">geom_histogram</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Histogram of Ex2"</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">"heart rate (bpm)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="lab1_intro_sol_files/figure-html/unnamed-chunk-2-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The graph is fairly normal-shaped, but with a longer right tail.</p>
<ol start="3" type="1">
<li>Compute the counts, means, and standard deviations of one of the numeric variables of your choice, separated by <code>Year</code> category.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># base R</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>year_list <span class="ot">=</span> <span class="fu">unique</span>(heart_rates<span class="sc">$</span>Year)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (year <span class="cf">in</span> year_list) {</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>  slice <span class="ot">&lt;-</span> heart_rates<span class="sc">$</span>Year <span class="sc">==</span> year</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"In year"</span>, year, <span class="st">"number of observations of Ex2 is:"</span>, <span class="fu">length</span>(heart_rates<span class="sc">$</span>Ex2[slice])))</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"In year"</span>, year, <span class="st">"mean value of Ex2 is:"</span>, <span class="fu">mean</span>(heart_rates<span class="sc">$</span>Ex2[slice])))</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"In year"</span>, year, <span class="st">"standard deviation of Ex2 is:"</span>, <span class="fu">sd</span>(heart_rates<span class="sc">$</span>Ex2[slice])))</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "In year 2017 number of observations of Ex2 is: 106"
[1] "In year 2017 mean value of Ex2 is: 104.650943396226"
[1] "In year 2017 standard deviation of Ex2 is: 24.0875555512564"
[1] "In year 2018 number of observations of Ex2 is: 137"
[1] "In year 2018 mean value of Ex2 is: 99.7956204379562"
[1] "In year 2018 standard deviation of Ex2 is: 19.9496198865938"
[1] "In year 2019 number of observations of Ex2 is: 170"
[1] "In year 2019 mean value of Ex2 is: 101.976470588235"
[1] "In year 2019 standard deviation of Ex2 is: 22.229292722108"
[1] "In year 2020 number of observations of Ex2 is: 172"
[1] "In year 2020 mean value of Ex2 is: 106.273255813953"
[1] "In year 2020 standard deviation of Ex2 is: 22.6945335635076"
[1] "In year 2021 number of observations of Ex2 is: 161"
[1] "In year 2021 mean value of Ex2 is: 108.049689440994"
[1] "In year 2021 standard deviation of Ex2 is: 19.844017121741"
[1] "In year 2022 number of observations of Ex2 is: 182"
[1] "In year 2022 mean value of Ex2 is: 109.21978021978"
[1] "In year 2022 standard deviation of Ex2 is: 24.8639623382085"
[1] "In year 2023 number of observations of Ex2 is: 140"
[1] "In year 2023 mean value of Ex2 is: 110.078571428571"
[1] "In year 2023 standard deviation of Ex2 is: 21.8068921377415"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># tidyverse</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>heart_rates <span class="sc">|&gt;</span> <span class="fu">group_by</span>(Year) <span class="sc">|&gt;</span> <span class="fu">summarise</span>(<span class="at">num =</span> <span class="fu">n</span>(), <span class="at">mean_ex2 =</span> <span class="fu">mean</span>(Ex2), <span class="at">sd_ex2 =</span> <span class="fu">sd</span>(Ex2))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 7 × 4
   Year   num mean_ex2 sd_ex2
  &lt;dbl&gt; &lt;int&gt;    &lt;dbl&gt;  &lt;dbl&gt;
1  2017   106    105.    24.1
2  2018   137     99.8   19.9
3  2019   170    102.    22.2
4  2020   172    106.    22.7
5  2021   161    108.    19.8
6  2022   182    109.    24.9
7  2023   140    110.    21.8</code></pre>
</div>
</div>
<ol start="4" type="1">
<li>Visualize one of the numeric variables as a box plot or violin plot, separated by different years.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># base R</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(Ex2 <span class="sc">~</span> Year, <span class="at">data =</span> heart_rates, <span class="at">main=</span> <span class="st">"Ex2 for different years"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab1_intro_sol_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ggplot</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>heart_rates <span class="sc">|&gt;</span> <span class="fu">ggplot</span>() <span class="sc">+</span> <span class="fu">aes</span>(<span class="at">x=</span><span class="fu">as.factor</span>(Year), <span class="at">y =</span> Ex2) <span class="sc">+</span> <span class="fu">geom_boxplot</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Ex2 for different years"</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">"year"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab1_intro_sol_files/figure-html/unnamed-chunk-4-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<ol start="5" type="1">
<li>Report if there are any missing values and if any points may be considered outliers.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># base R to calculate by column</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"The number of NAs in each column: "</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>The number of NAs in each column: </code></pre>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="fu">colSums</span>(<span class="fu">is.na</span>(heart_rates)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>0 0 0 0 0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># tidyverse to calculate by row</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>heart_rates <span class="sc">%&gt;%</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rowwise</span>() <span class="sc">%&gt;%</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">sum_na =</span> <span class="fu">sum</span>(<span class="fu">is.na</span>(<span class="fu">c_across</span>())))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: There was 1 warning in `mutate()`.
ℹ In argument: `sum_na = sum(is.na(c_across()))`.
ℹ In row 1.
Caused by warning:
! Using `c_across()` without supplying `cols` was deprecated in dplyr 1.1.0.
ℹ Please supply `cols` instead.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1,068 × 6
# Rowwise: 
   Rest1   Ex1 Rest2   Ex2  Year sum_na
   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;int&gt;
 1    86   103    84   112  2017      0
 2    68   108    68   128  2017      0
 3    71    88    82    96  2017      0
 4    64    75    66    78  2017      0
 5    78   119   109   111  2017      0
 6    60    89    60    87  2017      0
 7    68    76    72    76  2017      0
 8    77    90    81    99  2017      0
 9    70    66    76    75  2017      0
10    65    63    83    80  2017      0
# ℹ 1,058 more rows</code></pre>
</div>
</div>
<p>In the histogram you can see some points that are</p>
<section id="viral-mutation-rates" class="level3">
<h3 class="anchored" data-anchor-id="viral-mutation-rates">Viral mutation rates</h3>
<p>The following data set comes from <a href="https://github.com/lauringlab/JVI_Gem_2018" class="uri">https://github.com/lauringlab/JVI_Gem_2018</a> and contains measurements of different viral species. The three numeric variables are evolutionary rate K, mutation rate mu, and genome size G. The categorical variables are fairly self-explanatory.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>viral_mut_rates <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"https://raw.githubusercontent.com/lauringlab/JVI_Gem_2018/master/Figure_1_mu_and_K_data.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 111 Columns: 11
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (7): group, family, virus, K reference (Sanjuan 2012), Full K reference,...
dbl (3): G, K, mu
lgl (1): Included in 2018 publication

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
</div>
<ol type="1">
<li><p>Report the number and types of variables and the number of observations</p>
<p>11 columns, of which 3 are numeric: G, K, mu, one logical, and the rest are categorical, 111 observations</p></li>
<li><p>Make a histogram of one of the numeric variables of your choice; describe its shape (e.g.&nbsp;normal, uniform, bimodal) and comment on any interesting features.</p></li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># base R:</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(viral_mut_rates<span class="sc">$</span>G, <span class="at">main =</span> <span class="st">'Histogram of gene size'</span>, <span class="at">xlab =</span> <span class="st">"size (Kbp?)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab1_intro_sol_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ggplot</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>viral_mut_rates <span class="sc">|&gt;</span> <span class="fu">ggplot</span>() <span class="sc">+</span> <span class="fu">aes</span>(<span class="at">x=</span>G) <span class="sc">+</span> <span class="fu">geom_histogram</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">'Histogram of gene size'</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">"size (Kbp?)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="lab1_intro_sol_files/figure-html/unnamed-chunk-7-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<ol start="3" type="1">
<li>Compute the counts, means, and standard deviations of one of the numeric variables of your choice, separated by <code>group</code> category.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># base R</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>group_list <span class="ot">=</span> <span class="fu">unique</span>(viral_mut_rates<span class="sc">$</span>group)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (g <span class="cf">in</span> group_list) {</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>  slice <span class="ot">&lt;-</span> viral_mut_rates<span class="sc">$</span>group <span class="sc">==</span> g</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"In group"</span>, g, <span class="st">"number of observations of G is:"</span>, <span class="fu">length</span>(viral_mut_rates<span class="sc">$</span>G[slice])))</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"In group"</span>, g, <span class="st">"mean value of G is:"</span>, <span class="fu">mean</span>(viral_mut_rates<span class="sc">$</span>G[slice])))</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"In group"</span>, g, <span class="st">"standard deviation of G is:"</span>, <span class="fu">sd</span>(viral_mut_rates<span class="sc">$</span>G[slice])))</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "In group dsDNA number of observations of G is: 14"
[1] "In group dsDNA mean value of G is: 134.102857142857"
[1] "In group dsDNA standard deviation of G is: 166.383421531232"
[1] "In group dsRNA number of observations of G is: 6"
[1] "In group dsRNA mean value of G is: 17.2666666666667"
[1] "In group dsRNA standard deviation of G is: 6.79719255771577"
[1] "In group retro number of observations of G is: 9"
[1] "In group retro mean value of G is: 7.96111111111111"
[1] "In group retro standard deviation of G is: 3.16570467844224"
[1] "In group ss_neg_RNA number of observations of G is: 15"
[1] "In group ss_neg_RNA mean value of G is: 14.2"
[1] "In group ss_neg_RNA standard deviation of G is: 2.48985082743066"
[1] "In group ss_pos_RNA number of observations of G is: 55"
[1] "In group ss_pos_RNA mean value of G is: 10.0505454545455"
[1] "In group ss_pos_RNA standard deviation of G is: 6.5925834924085"
[1] "In group ssDNA number of observations of G is: 12"
[1] "In group ssDNA mean value of G is: 4.68"
[1] "In group ssDNA standard deviation of G is: 1.52453271529344"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># tidyverse</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>viral_mut_rates <span class="sc">|&gt;</span> <span class="fu">group_by</span>(group) <span class="sc">|&gt;</span> <span class="fu">summarise</span>(<span class="at">num =</span> <span class="fu">n</span>(), <span class="at">mean_G =</span> <span class="fu">mean</span>(G), <span class="at">sd_G =</span> <span class="fu">sd</span>(G))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
  group        num mean_G   sd_G
  &lt;chr&gt;      &lt;int&gt;  &lt;dbl&gt;  &lt;dbl&gt;
1 dsDNA         14 134.   166.  
2 dsRNA          6  17.3    6.80
3 retro          9   7.96   3.17
4 ssDNA         12   4.68   1.52
5 ss_neg_RNA    15  14.2    2.49
6 ss_pos_RNA    55  10.1    6.59</code></pre>
</div>
</div>
<ol start="4" type="1">
<li>Visualize one of the numeric variables as a box plot or violin plot, separated by group.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># base R</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(G <span class="sc">~</span> group, <span class="at">data =</span> viral_mut_rates, <span class="at">main=</span> <span class="st">"Genome size for different groups of viruses"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab1_intro_sol_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ggplot</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>viral_mut_rates <span class="sc">|&gt;</span> <span class="fu">ggplot</span>() <span class="sc">+</span> <span class="fu">aes</span>(<span class="at">x=</span>group, <span class="at">y =</span> G) <span class="sc">+</span> <span class="fu">geom_boxplot</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Genome size for different groups of viruses"</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">"virus group"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab1_intro_sol_files/figure-html/unnamed-chunk-9-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<ol start="5" type="1">
<li>Report if there are any missing values and if any points may be considered outliers.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># base R to calculate by column</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"The number of NAs in each column: "</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>The number of NAs in each column: </code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="fu">colSums</span>(<span class="fu">is.na</span>(viral_mut_rates)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>0 0 0 0 17 84 26 17 86 84 0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># tidyverse to calculate by row</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>viral_mut_rates <span class="sc">|&gt;</span> </span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(G,K, mu) <span class="sc">|&gt;</span> </span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rowwise</span>() <span class="sc">|&gt;</span> </span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">sum_na =</span> <span class="fu">sum</span>(<span class="fu">is.na</span>(<span class="fu">c_across</span>())))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 111 × 4
# Rowwise: 
        G           K           mu sum_na
    &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;  &lt;int&gt;
 1  35.9  NA           0.000000131      1
 2  35.3   0.000072   NA                1
 3  35.9   0.0000346  NA                1
 4 152.    0.0000821   0.000000059      0
 5 125.    0.00000626 NA                1
 6 235    NA           0.0000002        1
 7 169    NA           0.000000098      1
 8   7.9   0.0039     NA                1
 9   5.15  0.0000434  NA                1
10   5.13  0.000017   NA                1
# ℹ 101 more rows</code></pre>
</div>
</div>
</section>
<section id="neuroblastoma-data" class="level3">
<h3 class="anchored" data-anchor-id="neuroblastoma-data">Neuroblastoma data</h3>
<p>The following data set contains 22 scores of different from patients with neuroblastoma (a type of cancer) derived from cell-free DNA samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>neuro_blast <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"https://raw.githubusercontent.com/dkon1/intro-ml-bio/main/labs/data/r2_gse62564_GSVA_Metadata_selected.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 498 Columns: 24
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (2): sample id, high_risk
dbl (22): MYCN, STC1, P4HA1, BHLHE40, HIF1A, ELVIDGE_HIF1A_TARGETS_UP, Quies...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
</div>
<ol type="1">
<li><p>Report the number and types of variables and the number of observations</p>
<p>There are 24 columns, of which 22 are numeric, and two are categorical: sample id is a unique id so it won’t be used in calculations, and high_risk is a logical variable (yes/no)</p></li>
<li><p>Compare the distributions of two or more of the numeric variables of your choice using boxplots or other visualizations; describe how the distributions differ.</p></li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># base R:</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(neuro_blast<span class="sc">$</span>MYCN, <span class="at">main =</span> <span class="st">'Histogram of MYCN'</span>, <span class="at">xlab =</span> <span class="st">"size (Kbp?)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab1_intro_sol_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ggplot</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>neuro_blast <span class="sc">|&gt;</span> <span class="fu">ggplot</span>() <span class="sc">+</span> <span class="fu">aes</span>(<span class="at">x=</span>MYCN) <span class="sc">+</span> <span class="fu">geom_histogram</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">'Histogram of MYCN'</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">"MYCN"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="lab1_intro_sol_files/figure-html/unnamed-chunk-12-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The histogram of MYCN is very non-normal, the biggest peak is at 0, with a long tail of higher values.</p>
<ol start="3" type="1">
<li>Compute the counts, means, and standard deviations of one of the numeric variables of your choice, separated by <code>high_risk</code> category.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># base R</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>risk_list <span class="ot">=</span> <span class="fu">unique</span>(neuro_blast<span class="sc">$</span>high_risk)</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (risk <span class="cf">in</span> risk_list) {</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>  slice <span class="ot">&lt;-</span> neuro_blast<span class="sc">$</span>high_risk <span class="sc">==</span> risk</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"In group"</span>, risk, <span class="st">"number of observations of G is:"</span>, <span class="fu">length</span>(neuro_blast<span class="sc">$</span>MYCN[slice])))</span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"In group"</span>, risk, <span class="st">"mean value of G is:"</span>, <span class="fu">mean</span>(neuro_blast<span class="sc">$</span>MYCN[slice])))</span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"In group"</span>, risk, <span class="st">"standard deviation of G is:"</span>, <span class="fu">sd</span>(neuro_blast<span class="sc">$</span>MYCN[slice])))</span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "In group yes number of observations of G is: 176"
[1] "In group yes mean value of G is: 550.603602272727"
[1] "In group yes standard deviation of G is: 616.077018482528"
[1] "In group no number of observations of G is: 322"
[1] "In group no mean value of G is: 56.7189099378882"
[1] "In group no standard deviation of G is: 54.1659434545841"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># tidyverse</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>neuro_blast <span class="sc">|&gt;</span> <span class="fu">group_by</span>(high_risk) <span class="sc">|&gt;</span> <span class="fu">summarise</span>(<span class="at">num =</span> <span class="fu">n</span>(), <span class="at">mean_MYCN =</span> <span class="fu">mean</span>(MYCN), <span class="at">sd_MYCN =</span> <span class="fu">sd</span>(MYCN))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 4
  high_risk   num mean_MYCN sd_MYCN
  &lt;chr&gt;     &lt;int&gt;     &lt;dbl&gt;   &lt;dbl&gt;
1 no          322      56.7    54.2
2 yes         176     551.    616. </code></pre>
</div>
</div>
<p>You can see major differences in mean values of MYCN between high and low risk groups.</p>
<ol start="4" type="1">
<li>Visualize one of the numeric variables as a box plot or violin plot, separated by high_risk category.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co"># base R</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(MYCN <span class="sc">~</span> high_risk, <span class="at">data =</span> neuro_blast, <span class="at">main=</span> <span class="st">"MYCN for different risk groups"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab1_intro_sol_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ggplot</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>neuro_blast <span class="sc">|&gt;</span> <span class="fu">ggplot</span>() <span class="sc">+</span> <span class="fu">aes</span>(<span class="at">x=</span>high_risk, <span class="at">y =</span> MYCN) <span class="sc">+</span> <span class="fu">geom_boxplot</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"MYCN for different risk groups"</span>) <span class="sc">+</span> <span class="fu">xlab</span>(<span class="st">"High risk status"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="lab1_intro_sol_files/figure-html/unnamed-chunk-14-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<ol start="5" type="1">
<li>Report if there are any missing values and if any points may be considered outliers.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co"># base R to calculate by column</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"The number of NAs in each column: "</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>The number of NAs in each column: </code></pre>
</div>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="fu">colSums</span>(<span class="fu">is.na</span>(neuro_blast)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co"># tidyverse to calculate by row</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>neuro_blast <span class="sc">|&gt;</span>  <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(<span class="st">`</span><span class="at">sample id</span><span class="st">`</span>, high_risk)) <span class="sc">|&gt;</span> </span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rowwise</span>() <span class="sc">|&gt;</span> </span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">sum_na =</span> <span class="fu">sum</span>(<span class="fu">is.na</span>(<span class="fu">c_across</span>())))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 498 × 23
# Rowwise: 
     MYCN  STC1 P4HA1 BHLHE40 HIF1A ELVIDGE_HIF1A_TARGETS_UP
    &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;                    &lt;dbl&gt;
 1   42.5  2.95 12.9     20.9  83.4                   0.0661
 2   25.5  3.09 10.9     14.1  86.4                   0.410 
 3  883.   6.02 12.7     22.3  77.6                   0.0522
 4  433.   3.09 19.7     33.5  95.2                   0.102 
 5  854.  10.9  22.7     31.2 103.                    0.273 
 6  111.   9.32 13.6     24.4 105.                   -0.251 
 7   63.6  6.56 13.5     91.4  95.7                   0.132 
 8 1558.  58.7   9.72    32.1 128.                    0.135 
 9  727.  16.1  23.9     51.1  88.7                   0.188 
10 1234.   4.24 23.3     24.2 127.                    0.210 
# ℹ 488 more rows
# ℹ 17 more variables: `Quiescence Down Cheung` &lt;dbl&gt;,
#   `T cell exhaustion` &lt;dbl&gt;, HALLMARK_INFLAMMATORY_RESPONSE &lt;dbl&gt;,
#   HALLMARK_HYPOXIA &lt;dbl&gt;, ST8SIA1 &lt;dbl&gt;, `Quiescence Up Cheung` &lt;dbl&gt;,
#   HALLMARK_TNFA_SIGNALING_VIA_NFKB &lt;dbl&gt;, HALLMARK_MITOTIC_SPINDLE &lt;dbl&gt;,
#   HALLMARK_DNA_REPAIR &lt;dbl&gt;, HALLMARK_G2M_CHECKPOINT &lt;dbl&gt;,
#   HALLMARK_APOPTOSIS &lt;dbl&gt;, HALLMARK_E2F_TARGETS &lt;dbl&gt;, …</code></pre>
</div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>