---
title: "Classification methods (lab 4 for BIOS 26122)"
author: "Dmitry Kondrashov"
format: 
  html:
    self-contained: true
editor: visual
---

## Description

The goal of this assignment is to perform classification tasks by training a model and then validating its performance on the test set. You will learn to use the following models:

1.  Generalized linear model

2.  Naive Bayes

3.  Linear Discriminant Analysis

4.  Quadratic Discriminant Analysis

The use of these models is demonstrated in the week 4 tutorials using the tools from package `tidymodels`; I recommend that you use them to perform the tasks below.

```{r setup}
#| include: false
#| echo: false
library(tidyverse)
library(tidymodels)
library(discrim)
```

### Neuroblastoma data

The following data set is gene expression data from tumors of patients with neuroblastoma (a type of cancer), accession number [**GSE62564**](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE62564). It contains 22 phenotypic scores, 6 of which (MYCN, STC1, P4HA1, BHLHE40, HIF1A and ST8SIA1) are gene expressions measured in log2RPM (log 2 reads per million). The other 16 are quantified by R package GSVA (Gene Set Enrichment Analysis).

```{r}
neuro_blast <- read_csv("https://raw.githubusercontent.com/dkon1/intro-ml-bio/main/labs/data/r2_gse62564_GSVA_Metadata_selected.csv")
```

1.  Clean the data to remove any outliers or missing values. Split the data set into training and test sets of equal size. Set up a tidymodels recipe to predict the variable `high_risk` (make sure to convert it to a factor!) using all other variables, except for `sample_id`

    ```{r}
    neuro_clean <- neuro_blast |>  drop_na() |>
      mutate(high_risk = factor(high_risk))

    # Put 1/2 of the data into the training set 
    neuro_split <- initial_split(neuro_clean, prop = 1/2)

    # Create data frames for the two sets:
    neuro_train <- training(neuro_split)
    neuro_test  <- testing(neuro_split)


    neuro_recipe <- 
      recipe(high_risk ~ ., data = neuro_train) |>  
      update_role(`sample id`, new_role = "ID")
    ```

2.  Train a *generalized linear model* (using function \`logistic_reg\` and engine "glm") to predict the variable `high_risk`; specify the model, create a workflow, fit it to the training set, print out the fitted parameters, and evaluate its performance on the test set.

```{r}
glm_spec <- 
    logistic_reg() |>  
    set_engine("glm")

workflow_glm <- workflow() |> 
  add_model(glm_spec) |> 
  add_recipe(neuro_recipe)

fit_neuro <- workflow_glm |> 
  fit(neuro_train)

#fit_neuro %>% 
 # extract_fit_parsnip() 

compare_pred <- augment(fit_neuro, new_data = neuro_test) 

compare_pred |>  accuracy(truth = high_risk, estimate = .pred_class)

compare_pred |>  conf_mat(truth = high_risk, estimate = .pred_class)
```

3.  Train a *Naive Bayes model* (using function `naive_Bayes` and engine "naivebayes") to predict the variable `high_risk`: specify the model, create a workflow, fit it to the training set, and report its accuracy on the test set.

    ```{r}
    nb_spec <- naive_Bayes() |>  
      set_mode("classification") |>  
      set_engine("naivebayes") |>  
      set_args(usekernel = FALSE) 


    workflow_nb <- workflow() |> 
      add_model(nb_spec) |> 
      add_recipe(neuro_recipe)

    fit_neuro <- workflow_nb |> 
      fit(neuro_train)

    #fit_neuro %>% 
     # extract_fit_parsnip() 

    compare_pred <- augment(fit_neuro, new_data = neuro_test) 

    compare_pred |>  accuracy(truth = high_risk, estimate = .pred_class)

    compare_pred |>  conf_mat(truth = high_risk, estimate = .pred_class)
    ```

4.  Train a *linear discriminant model* (using function `discrim_linear` with engine "MASS") to predict the variable `high_risk`: specify the model, create a workflow, fit it to the training set, and report its accuracy on the test set.

    ```{r}
    ld_spec <- discrim_linear() |> 
      set_mode("classification") |>  
      set_engine("MASS") |>  
      set_args(penalty = NULL, regularization_method = NULL) 


    workflow_ld <- workflow() |> 
      add_model(ld_spec) |> 
      add_recipe(neuro_recipe)

    fit_neuro <- workflow_ld |> 
      fit(neuro_train)

    #fit_neuro %>% 
     # extract_fit_parsnip() 

    compare_pred <- augment(fit_neuro, new_data = neuro_test) 

    compare_pred |>  accuracy(truth = high_risk, estimate = .pred_class)

    compare_pred |>  conf_mat(truth = high_risk, estimate = .pred_class)
    ```

5.  Train a *quadratic discriminant model* (using function `discrim_quad` with engine "MASS") to predict the variable `high_risk`: specify the model, create a workflow, fit it to the training set, and report its accuracy on the test set.

```{r}
 qd_spec <- discrim_quad()  |> 
  set_mode("classification") |>  
  set_engine("MASS") |>  
  set_args(penalty = NULL, regularization_method = NULL) 


workflow_qd <- workflow() |> 
  add_model(qd_spec) |> 
  add_recipe(neuro_recipe)

fit_neuro <- workflow_qd |> 
  fit(neuro_train)

#fit_neuro %>% 
 # extract_fit_parsnip() 

compare_pred <- augment(fit_neuro, new_data = neuro_test) 

compare_pred |>  accuracy(truth = high_risk, estimate = .pred_class)

compare_pred |>  conf_mat(truth = high_risk, estimate = .pred_class)
```

### Ecological data

The following data set contains observations of the populations of one species of fish (cutthroat trout) and two species of salamander in Mack Creek, Andrews Forest, Willamette National Forest, Oregon. The data set contains 16 variables and over thirty-two thousand observations. The variables include time and date, location, and measurements, such as size and weight. The metadata (descriptions of data) are provided [here](https://portal.edirepository.org/nis/metadataviewer?packageid=knb-lter-and.4027.14) (click on "Data entities" tab for explanations of each variable.)

```{r}
mack_data <- read_csv("https://raw.githubusercontent.com/dkon1/quant_life_quarto/main/data/mack_data.csv")
```

1.  Clean the data to remove any outliers or missing values, and filter the data to contain observations from only one species, either: 'Cutthroat trout' or 'Coastal giant salamander' . Split the data set into training and test sets of equal size. Set up a `tidymodels` recipe to predict the variable \`section\` (make sure it's converted to a factor) using the numeric variables `length_1_mm` and `weight_g`.

    ```{r}
    mack_clean <- mack_data |> 
      filter(species == 'Coastal giant salamander') |>  
      dplyr::select(section, length_1_mm, weight_g) |> 
      drop_na() |> mutate(section = factor(section))

    # Put 3/4 of the data into the training set 
    mack_split <- initial_split(mack_clean, prop = 1/2)

    # Create data frames for the two sets:
    mack_train <- training(mack_split)
    mack_test  <- testing(mack_split)


    mack_recipe <- 
      recipe(section ~ ., data = mack_train)
    ```

2.  Train a *generalized linear model* (using engine "glm") to predict the variable `section`; specify the model, create a workflow, fit it to the training set, print out the fitted parameters, and evaluate its performance on the test set.

```{r}
glm_spec <- logistic_reg() |>  
    set_engine("glm")

workflow_glm <- workflow() |> 
  add_model(glm_spec) |> 
  add_recipe(mack_recipe)

fit_mack <- workflow_glm |> 
  fit(mack_train)

#fit_mack %>% 
 # extract_fit_parsnip() 

compare_pred <- augment(fit_mack, new_data = mack_test) 

compare_pred |>  accuracy(truth = section, estimate = .pred_class)

compare_pred |>  conf_mat(truth = section, estimate = .pred_class)
```

3.  Train a *Naive Bayes model* to predict the variable `section`: specify the model, create a workflow, fit it to the training set, print out the fitted parameters, and evaluate its performance on the test set.

    ```{r}
    nb_spec <- naive_Bayes() %>% 
      set_mode("classification") %>% 
      set_engine("naivebayes") %>% 
      set_args(usekernel = FALSE) 


    workflow_nb <- workflow() |> 
      add_model(nb_spec) |> 
      add_recipe(mack_recipe)

    fit_mack <- workflow_nb |> 
      fit(mack_train)

    #fit_mack %>% 
     # extract_fit_parsnip() 

    compare_pred <- augment(fit_mack, new_data = mack_test) 

    compare_pred |>  accuracy(truth = section, estimate = .pred_class)

    compare_pred |>  conf_mat(truth = section, estimate = .pred_class)
    ```

4.  Train a *linear discriminant model* to predict the variable `section`: specify the model, create a workflow, fit it to the training set, print out the fitted parameters, and evaluate its performance on the test set.

    ```{r}
    ld_spec <- discrim_linear(
      mode = "classification",
      penalty = NULL,
      regularization_method = NULL,
      engine = "MASS"
    )

    workflow_ld <- workflow() |> 
      add_model(ld_spec) |> 
      add_recipe(mack_recipe)

    fit_mack <- workflow_ld |> 
      fit(mack_train)

    #fit_mack %>% 
     # extract_fit_parsnip() 

    compare_pred <- augment(fit_mack, new_data = mack_test) 

    compare_pred |>  accuracy(truth = section, estimate = .pred_class)

    compare_pred |>  conf_mat(truth = section, estimate = .pred_class)
    ```

5.  Train a *quadratic discriminant model* to predict the variable `section`: specify the model, create a workflow, fit it to the training set, and report its accuracy on the test set.

```{r}
 qd_spec <- discrim_quad(
  mode = "classification",
  regularization_method = NULL,
  engine = "MASS"
)

workflow_qd <- workflow() |> 
  add_model(qd_spec) |> 
  add_recipe(mack_recipe)

fit_mack <- workflow_qd |> 
  fit(mack_train)

#fit_neuro %>% 
 # extract_fit_parsnip() 

compare_pred <- augment(fit_mack, new_data = mack_test) 

compare_pred |>  accuracy(truth = section, estimate = .pred_class)

compare_pred |>  conf_mat(truth = section, estimate = .pred_class)
```

6.  (Optional) Go back to question 1 and add one or a few more predictor variables to the recipe for prediction of section (think about which one make sense to use as predictors) and report the results of fitting this model using your method of choice.

    ```{r}

    ```

```{r}
yeast_data <- read_csv("yeast.csv")
```

```{r}
yeast_data <- yeast_data |> 
  filter(name %in% c("NUC", "CYT")) |> 
  mutate(name = as.factor(name))  
   #|>  
#  dplyr::select(section, length_1_mm, weight_g) |> 
#  drop_na() |> mutate(section = factor(section))

# Put 3/4 of the data into the training set 
yeast_split <- initial_split(yeast_data, prop = 3/4)

# Create data frames for the two sets:
yeast_train <- training(yeast_split)
yeast_test  <- testing(yeast_split)


yeast_recipe <- 
  recipe(name ~ ., data = yeast_train)
```

Variable names:

1.  Sequence: Accession number for the SWISS-PROT database
2.  mcg: McGeoch's method for signal sequence recognition.
3.  gvh: von Heijne's method for signal sequence recognition.
4.  lip: von Heijne's Signal Peptidase II consensus sequence score.
5.  chg: Presence of charge on N-terminus of predicted lipoproteins.
6.  aac: score of discriminant analysis of the amino acid content of outer membrane and periplasmic proteins.
7.  alm1: score of the ALOM membrane spanning region prediction program.
8.  alm2: score of ALOM program after excluding putative cleavable signal regions from the sequence.
9.  Class, or the localization site is one of the following categories: cp (cytoplasm); im (inner membrane without signal sequence); pp (perisplasm); imU (inner membrane, uncleavable signal sequence); om (outer membrane); omL (outer membrane lipoprotein); imL (inner membrane lipoprotein); imS (inner membrane, cleavable signal sequence)

```{r}
ecoli_data <- read_csv("ecoli.csv")
```

```{r}
ecoli_data <- ecoli_data  |> 
 # filter(name %in% c("NUC", "CYT")) |> 
  dplyr::select(-lip, -chg) |> 
  filter(class %in% c("cp","im","pp","om","imU")) |> 
  mutate(class= as.factor(class))  
   #|>  
#  dplyr::select(section, length_1_mm, weight_g) |> 
#  drop_na() |> mutate(section = factor(section))

# Put 3/4 of the data into the training set 
ecoli_split <- initial_split(ecoli_data, prop = 0.5)

# Create data frames for the two sets:
ecoli_train <- training(ecoli_split)
ecoli_test  <- testing(ecoli_split)


ecoli_recipe <- 
  recipe(class ~ ., data = ecoli_train) |> 
  update_role(Sequence, new_role = "ID") 
```

```{r}
nb_spec <- naive_Bayes() %>% 
  set_mode("classification") %>% 
  set_engine("naivebayes") %>% 
  set_args(usekernel = FALSE) 


workflow_nb <- workflow() |> 
  add_model(nb_spec) |> 
  add_recipe(ecoli_recipe)

fit_ecoli <- workflow_nb |> 
  fit(ecoli_train)

#fit_mack %>% 
 # extract_fit_parsnip() 

compare_pred <- augment(fit_ecoli, new_data = ecoli_test) 

compare_pred |>  accuracy(truth = class, estimate = .pred_class)

compare_pred |>  conf_mat(truth = class, estimate = .pred_class)
```

```{r}
ld_spec <- discrim_linear(
  mode = "classification",
  penalty = NULL,
  regularization_method = NULL,
  engine = "MASS"
)

workflow_ld <- workflow() |> 
  add_model(ld_spec) |> 
  add_recipe(ecoli_recipe)

fit_ecoli <- workflow_ld |> 
  fit(ecoli_train)

#fit_mack %>% 
 # extract_fit_parsnip() 

compare_pred <- augment(fit_ecoli, new_data = ecoli_test) 

compare_pred |>  accuracy(truth = class, estimate = .pred_class)

compare_pred |>  conf_mat(truth = class, estimate = .pred_class)
```

Notes: this dataset present some problems for classification because some of the predictor variables don't vary! The other problem is that some of the classes are too rare and this makes NB break when there is none in the trainin set. This is a good example of needing to clean the dataset to remove some variables (and potentially simplify the classes) to improve prediction.

```{r}

```
